src/data/int/modeq.lean

```diff
@@ -146,13 +155,40 @@ by rw [modeq_iff_dvd] at *; exact (dvd_mul_left n m).trans h
 theorem of_mul_right (m : ℤ) : a ≡ b [ZMOD n * m] → a ≡ b [ZMOD n] :=
 mul_comm m n ▸ of_mul_left _
 
+/-- To cancel a common factor `c` from a `modeq` we must divide the modulus `m` by `gcd m c`. -/
+lemma cancel_left_div_gcd (hm : 0 < m) (h : c * a ≡ c * b [ZMOD m]) : a ≡ b [ZMOD m / gcd m c] :=
+begin
+  let d := gcd m c,
+  have hmd := gcd_dvd_left m c,
+  have hcd := gcd_dvd_right m c,
+  rw modeq_iff_dvd at ⊢ h,
+  refine int.dvd_of_dvd_mul_right_of_gcd_one _ _,
+  show m / d ∣ c / d * (b - a),
+  { rw [mul_comm, ←int.mul_div_assoc (b - a) hcd, mul_comm],
+    apply int.div_dvd_div hmd,
+    rwa mul_sub },
+  { rw [gcd_div hmd hcd, nat_abs_of_nat, nat.div_self (gcd_pos_of_ne_zero_left c hm.ne')] }
+end
+
+lemma cancel_right_div_gcd (hm : 0 < m) (h : a * c ≡ b * c [ZMOD m]) : a ≡ b [ZMOD m / gcd m c] :=
+by { apply cancel_left_div_gcd hm, simpa [mul_comm] using h }
+
+lemma of_div (h : a / c ≡ b / c [ZMOD m / c]) (ha : c ∣ a) (ha : c ∣ b) (ha : c ∣ m) :
+  a ≡ b [ZMOD m] :=
+by convert h.mul_left'; rwa int.mul_div_cancel'
+
 end modeq
 
 theorem modeq_one : a ≡ b [ZMOD 1] := modeq_of_dvd (one_dvd _)
 
 lemma modeq_sub (a b : ℤ) : a ≡ b [ZMOD a - b] :=
 (modeq_of_dvd dvd_rfl).symm
 
+@[simp] lemma modeq_zero_iff : a ≡ b [ZMOD 0] ↔ a = b := by rw [modeq, mod_zero, mod_zero]
```

Comment:
If you wanted you could add the even stronger statement that `int.zmodeq 0 = eq` or whatever the legal spelling is

Reply:
Those underapplied lemmas are usually pretty useless because you can't finely rewrite with them.